Lib/socketserver.py

```diff
@@ -651,24 +652,46 @@ def process_request_thread(self, request, client_address):
         except Exception:
             self.handle_error(request, client_address)
         finally:
-            self.shutdown_request(request)
+            try:
+                self.shutdown_request(request)
+            finally:
+                self._remove_thread()
+
+    def _remove_thread(self):
+        """Remove a current thread from threads list."""
+        thread = threading.current_thread()
+        if not thread.daemon:
+            if self._threads_lock is None:
+                self._threads_lock = threading.Lock()
```

Comment:
That's wrong: the lock must exist before the first thread is spawned. Otherwise, two threads might use two different locks... (There would be a race condition on the lock creation.)

Reply:
The lock won't have been initialized if _block_on_close_ was false.

Reply:
The lock is now created with the list of threads as a `_Threads` object. The `block_on_close` is used to determine the type of object to create for `_threads` so that it's referenced exactly once and has no effect after the first request is handled.